### PR TITLE
ENH: vendor typing_extensions

### LINKED ISSUE #34869: ENH: Vendor typing_extensions
I’ve discussed this with @simonjayhawkins here and there but never had a formal issue opened, hence this :-)

I think we need to vendor typing_extensions to realistically be able to unpin mypy from where it is now.  The main thing I think we need is to unlock Protocols; these aren’t added until Python 3.8 but as of mypy 0.75 seem to be the documented solution to a lot of the Mixin problems that we face:

https://mypy-lang.blogspot.com/2019/11/mypy-0.html

There are some other useful things like Literal, Final and TypedDict that this would unlock as well, to at least hold us over until 3.8 is the minimum supported version

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #34869
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

haven't figured out how to make isort and black ignore this file

### CODE REVIEW

**Comment by jbrockmendel** [comment]
cc @WillAyd @simonjayhawkins thoughts on getting this to pass checks?

**Comment by WillAyd** [comment]
FWIW I think OK to run the new module through isort / black; doesn't need to be a perfect reproduction

AFAICT code check failures are mostly around internal style rules we place; OK to update to those as well as they are mostly cosmetic

**Review by simonjayhawkins — APPROVED** [review]
Thanks @jbrockmendel 

do we need any sort of testing for this?

File: setup.cfg [hunk 1]
```diff
@@ -29,6 +29,7 @@ exclude =
     doc/build/*.py,
     doc/temp/*.py,
     .eggs/*.py,
+    pandas/_vendored/*
```
Comment (WillAyd) [inline setup.cfg:32]: Is this still required to ignore?

Comment (jbrockmendel) [reply setup.cfg:32]: either way i guess

**Comment by jbrockmendel** [comment]
> do we need any sort of testing for this?

I wouldn't bother.  FWIW not needing to worry about that is part of why i suggested making this a dependency in #34869

File: setup.cfg (hunk 1 above)
Comment (WillAyd) [reply setup.cfg:32]: Let's remove if not necessary; less things to ponder is always better

**Review by WillAyd — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jbrockmendel

**Pull request merged**